src/web/app/components/course-edit-form/course-edit-form.component.spec.ts

```diff
@@ -1,35 +1,177 @@
+} from '../../../types/api-output';
+import { AjaxLoadingModule } from '../ajax-loading/ajax-loading.module';
+import { CopyCourseModalComponent } from '../copy-course-modal/copy-course-modal.component';
+import { LoadingRetryModule } from '../loading-retry/loading-retry.module';
+import { LoadingSpinnerModule } from '../loading-spinner/loading-spinner.module';
+import {
+  CourseAddFormModel,
+   CourseEditFormMode,
+   DEFAULT_COURSE_ADD_FORM_MODEL,
+} from './course-edit-form-model';
 import { CourseEditFormComponent } from './course-edit-form.component';
 
 describe('CourseEditFormComponent', () => {
   let component: CourseEditFormComponent;
   let fixture: ComponentFixture<CourseEditFormComponent>;
+  let feedbackSessionsService: FeedbackSessionsService;
+  let ngbModal: NgbModal;
 
-  beforeEach(async () => {
-    await TestBed.configureTestingModule({
-      declarations: [
-        CourseEditFormComponent,
-        AjaxLoadingComponent,
-      ],
+  const validCourseId: string = 'CS1101S';
+  const validTimeZone: string = 'Asia/Singapore';
+  const validInstitute: string = 'Test Institute';
+  const testTimeZone: string = 'Australia/Adelaide';
+  const timeZoneOffsets1: Record<string, number> = { GMT: 570 };
+  const testCourse: Course = {
+    courseId: 'testId',
+    courseName: 'Test Course',
+    timeZone: validTimeZone,
+    institute: validInstitute,
+    creationTimestamp: 0,
+    deletionTimestamp: 1000,
+  };
+
+  const spyStatusMessageService: any = {
+    showErrorToast: jest.fn(),
+    showSuccessToast: jest.fn(),
+  };
+  const timezoneServiceStub: any = {
+    getTzOffsets: jest.fn(() => timeZoneOffsets1),
+    guessTimezone: jest.fn(() => testTimeZone),
+  };
+
+  const spyCourseService: any = {
+    createCourse: jest.fn(() => of({})),
+    getAllCoursesAsInstructor: jest.fn(() => of({ courses: [testCourse] })),
+  };
+
+  beforeEach(waitForAsync(() => {
+    TestBed.configureTestingModule({
+      declarations: [CourseEditFormComponent],
       imports: [
         HttpClientTestingModule,
         FormsModule,
+        ReactiveFormsModule,
+        RouterTestingModule,
+        NgbModule,
+        AjaxLoadingModule,
+        CommonModule,
+        FormsModule,
+        AjaxLoadingModule,
+        LoadingRetryModule,
+        LoadingSpinnerModule,
+      ],
+      providers: [
+        { provide: StatusMessageService, useValue: spyStatusMessageService },
+        { provide: CourseService, useValue: spyCourseService },
+        { provide: TimezoneService, useValue: timezoneServiceStub },
       ],
     })
-      .compileComponents();
-  });
+    .compileComponents();
+  }));
 
   beforeEach(() => {
     fixture = TestBed.createComponent(CourseEditFormComponent);
+    feedbackSessionsService = TestBed.inject(FeedbackSessionsService);
+    ngbModal = TestBed.inject(NgbModal);
     component = fixture.componentInstance;
-    fixture.detectChanges();
   });
 
   it('should create', () => {
+    fixture.detectChanges();
```

Comment:
It is alright for `fixture.detectChanges` to be called multiple times; I think it would be better to keep the `fixture.detectChanges` in `beforeEach` instead of shifting it here, for consistency with all the other frontend tests.

Reply:
Understood! Will update!

Reply:
@zhaojj2209 I think I have to use `fixture.detectChanges` under each teat since data binding is done when this method called. I have not specified all the data eg. `CourseEditFormMode` in `beforeEach`, so using `fixture.detectChanges` in `beforeEach` would give weird behavior. May I have your thoughts on this?

Reply:
It's alright to have multiple calls of `fixture.detectChanges`, so even after calling `fixture.detectChanges` in `beforeEach`, you can still call it again after modifying various parameters. In other words, there was no need to remove `fixture.detectChanges` from `beforeEach` even when calling `fixture.detectChanges` in individual tests.

All that's left is to add `fixture.detectChanges` back in `beforeEach`; everything else looks good 👍

Reply:
@zhaojj2209 I see, thank you very much for sharing this. For some reasons, my tests fail when I add `fixture.detectChanges` in `beforeEach`. Any advice why this might be?
These are some of the errors I get:
![Screenshot from 2022-10-18 17-50-17](https://user-images.githubusercontent.com/59680441/196398126-392d008a-9542-4e73-b3a0-4dcfaf218ae7.png)
![image](https://user-images.githubusercontent.com/59680441/196397811-e9d74cd8-c8f7-4442-953b-5a036be93439.png)

Reply:
While it's understandable that the snapshot tests might fail since `fixture.detectChanges()` caused differences in the snapshots generated, what I didn't expect was for the extra `fixture.detectChanges()` call to break the modal mocks.

Seems that removing `fixture.detectChanges()` from `beforeEach` really is necessary for this testcase. Sorry for the confusion caused by my suggestion, no need for further changes on this!

Reply:
@zhaojj2209 Thank you for the clarification.